test/forge/HalmosTest.sol

```diff
@@ -0,0 +1,151 @@
+        vm.stopPrank();
+
+        vm.roll(block.number + 1);
+        vm.warp(block.timestamp + 1 * BLOCK_TIME);
+    }
+
+    // Call Morpho, assuming interacting with only the defined market for performance reasons.
+    function _callMorpho(bytes4 selector, address caller) internal {
+        vm.assume(selector != morpho.extSloads.selector);
+        vm.assume(selector != morpho.createMarket.selector);
+
+        bytes memory emptyData = hex"";
+        uint256 assets = svm.createUint256("assets");
+        uint256 shares = svm.createUint256("shares");
+        address onBehalf = svm.createAddress("onBehalf");
+        address receiver = svm.createAddress("receiver");
+
+        bytes memory args;
+
+        if (selector == morpho.supply.selector) {
+            args = abi.encode(marketParams, assets, shares, onBehalf, emptyData);
+        } else if (selector == morpho.withdraw.selector) {
+            args = abi.encode(marketParams, assets, shares, onBehalf, receiver);
+        } else if (selector == morpho.borrow.selector) {
+            args = abi.encode(marketParams, assets, shares, onBehalf, receiver);
+        } else if (selector == morpho.repay.selector) {
+            args = abi.encode(marketParams, assets, shares, onBehalf, emptyData);
+        } else if (selector == morpho.supplyCollateral.selector) {
+            args = abi.encode(marketParams, assets, onBehalf, emptyData);
+        } else if (selector == morpho.withdrawCollateral.selector) {
+            args = abi.encode(marketParams, assets, onBehalf, receiver);
+        } else if (selector == morpho.liquidate.selector) {
+            address borrower = svm.createAddress("borrower");
+            args = abi.encode(marketParams, borrower, assets, shares, emptyData);
+        } else if (selector == morpho.flashLoan.selector) {
+            uint256 rand = svm.createUint256("rand");
+            address token;
+            if (rand == 0) {
+                token = address(loanToken);
+            } else if (rand == 1) {
+                token = address(collateralToken);
+            } else {
+                ERC20Mock otherToken = new ERC20Mock();
+                token = address(otherToken);
+            }
+            args = abi.encode(marketParams, token, assets, emptyData);
+        } else if (selector == morpho.accrueInterest.selector) {
+            args = abi.encode(marketParams);
+        } else if (selector == morpho.setFee.selector) {
+            uint256 newFee = svm.createUint256("newFee");
+            args = abi.encode(marketParams, newFee);
+        } else {
+            args = svm.createBytes(1024, "data");
+        }
+
+        vm.prank(caller);
+        (bool success,) = address(morpho).call(abi.encodePacked(selector, args));
+        vm.assume(success);
+    }
+
+    // Check that the fee is always smaller than the max fee.
+    function check_feeInRange(bytes4 selector, address caller) public {
+        _callMorpho(selector, caller);
```

Comment:
I have not tried to do a loop, but here the whole state is symbolic, see the `--symbolic-storage` flag at the top of the test contract definition. It means that it proves the properties by starting in an arbitrary state, and in theory it is enough to show that this is an invariant. It's not entirely true though, because there are some variables that are not symbolic, such as the market that is created in the setup. But it should still give us a very good verification work for those properties

Reply:
ok, I completely missed this. Let me update my checks then

Reply:
why did they write this article then ?
>Have you tried just looping like they do [here](https://a16zcrypto.com/posts/article/implementing-stateful-invariant-testing-with-halmos/) ?

Reply:
> why did they write this article then ?

I'm assuming it's because they where talking about the usual case where you don't give the flag `--symbolic-storage`. Also there is the caveat that I mentioned about the storage not being completely symbolic